sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java

```diff
@@ -838,6 +843,31 @@ static ParsedNodes create(final SqlNode node) throws ValidationException
       return new ParsedNodes(explain, druidSqlInsert, query, ingestionGranularity);
     }
 
+    private static void validateSelectColumnForIngestion(SqlNode sqlNode) throws ValidationException
+    {
+      if (sqlNode instanceof SqlBasicCall) {
+        // If the sql node is a basic call, it could have select statements as operands.
+        // We need to recursivly check the operands for select statements.
+        SqlBasicCall sqlBasicCall = (SqlBasicCall) sqlNode;
+        for (SqlNode operand : sqlBasicCall.getOperandList()) {
+          validateSelectColumnForIngestion(operand);
+        }
+      } else if (sqlNode instanceof SqlSelect) {
+        // Validate that all columns which do some computation in the select statement are named.
```

Comment:
What would happen with a query like this?

```
INSERT INTO tbl
SELECT TIME_PARSE(timestamp) AS __time, *
FROM (SELECT timestamp, LOWER(x) FROM tbl)
PARTITIONED BY DAY
```

This query should be flagged as invalid. But I'm wondering if the structure would trick the validator, since it looks like the `*` is fine, but it's really going to get a funky name from `LOWER(x)`.

Reply:
One idea for handling such cases : can we use the `rootQueryRel` object after creating the `ParsedNodes` to see that there aren't any generated aliases? The `rootQueryRel` contains a `field` list which has ordinal to alias mapping from what I understand.
And for checking if an alias is generated by calcite we could use something like https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/sql/SqlUtil.java#L833 (directly not present in the calcite version we use).
The downside of doing this is that users would not be able to provide column names like `EXPR$<number>` - not sure if we can live with that.